File: matlab/PCFICH/CFICH/cfi_decoding.m
% /* 
%  * Copyright (c) 2012-2013, Ismael Gomez-Miguelez <ismael.gomez@tsc.upc.edu>.
%  * This file is part of ALOE++ (http://flexnets.upc.edu/)
%  * 
%  * ALOE++ is free software: you can redistribute it and/or modify
%  * it under the terms of the GNU Lesser General Public License as published by
%  * the Free Software Foundation, either version 3 of the License, or
%  * (at your option) any later version.
%  * 
%  * ALOE++ is distributed in the hope that it will be useful,
%  * but WITHOUT ANY WARRANTY; without even the implied warranty of
%  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
%  * GNU Lesser General Public License for more details.
%  * 
%  * You should have received a copy of the GNU Lesser General Public License
%  * along with ALOE++.  If not, see <http://www.gnu.org/licenses/>.
%  */

% Control Format Indicator (CFI) decoding. Determines the CFI value from a
% 32-bit sequence of.

%% Parameters
% Inputs:
% input         - 32-bit sequence at the receiver
%
%
% Outputs:
% out           - CFI value estimate
%

% Spec:         (3GPP TS 36.212, v10.6.0 section 5.3.4.1)
% Notes:        -
% Author:       Vuk Marojevic (vuk.marojevic@gmail.com)
% Last Revision:18.2.2013
% 


function out = cfi_decoding(input)

    seq = zeros(4,32);
    seq(1,:) = [0,1,1,0,1,1,0,1,1,0,1,1,0,1,1,0,1,1,0,1,1,0,1,1,0,1,1,0,1,1,0,1];
    seq(2,:) = [1,0,1,1,0,1,1,0,1,1,0,1,1,0,1,1,0,1,1,0,1,1,0,1,1,0,1,1,0,1,1,0];
    seq(3,:) = [1,1,0,1,1,0,1,1,0,1,1,0,1,1,0,1,1,0,1,1,0,1,1,0,1,1,0,1,1,0,1,1];
    seq(4,:) = [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0];
    % 4th sequence may not be necessary
    
    min = 32;
    for i=1:4
        ones_count = sum(mod(input+seq(i,:), 2));
        if (ones_count < min)
            min = ones_count;
            out = i;
        end
    end
end

File: modrep_osld/lte_cfi_decoding/src/cfi_decoding.c
/* 
 * Copyright (c) 2012, Ismael Gomez-Miguelez <ismael.gomez@tsc.upc.edu>.
 * This file is part of ALOE++ (http://flexnets.upc.edu/)
 * 
 * ALOE++ is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * ALOE++ is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public License
 * along with ALOE++.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <stdio.h>
#include <oesr.h>
#include <params.h>
#include <skeleton.h>

#define INCLUDE_DEFS_ONLY
#include "cfi_decoding.h"



/**
 * @ingroup lte_cfi_decoding
 * Decoding function
 * Decodes the recevied bitsequence to obtain the most likely CFI
 * The function performs bitwise XOR with each ot the coding sequences and
 * sums up the output. The minimum results indicates the code index and CFI.
 *
 * @param in Input bit sequence, received from descrambler, 32 bits
 * @param table Coding table, 4x32 bits
 * \returns Coding table index most similar to the recevied input sequence
 */
int cfi_decoding(char *in, char (*table)[NOF_BITS]) {

	int i, j;
	int count_i, index;
	int min = 32;

	for (i=0;i<4;i++) {
		count_i = 0;
		for (j=0;j<NOF_BITS;j++) {
			if (in[j] != table[i][j]) {
				count_i++;
			}
		}
		if (count_i < min) {
			min = count_i;
			index = i;
		}
	}
	return index;
}
